File: 380/insert-delete-getrandom-o1.java
class RandomizedSet {
    ArrayList<Integer> list = new ArrayList<Integer>();
    HashMap<Integer,Integer> map = new HashMap<>();
    public RandomizedSet() {
        
    }
    
    public boolean insert(int val) {
        if(!map.containsKey(val)){
            list.add(val);
            map.put(val, list.size()-1);
            return true;
        }
        return false;
    }
    
    public boolean remove(int val) {
        if(map.containsKey(val)){
            int i = map.get(val);
            map.remove(val);
            if(list.size()>1 && i!=list.size()-1){//這樣就避開問題了！
                map.put(list.get(list.size()-1), i);
                list.set(i, list.get(list.size()-1));
            }
            list.remove(list.size()-1);
            return true;
        }
        return false;
    }
    
    public int getRandom() {
        int i = (int)Math.floor(Math.random()*list.size());
        return list.get(i);
    }
}
//case 4/19: ["RandomizedSet","remove","remove","insert","getRandom","remove","insert"]
//[[],[0],[0],[0],[],[0],[0]]
//要確認在交換時，如果只剩1個時，是不能再map.put()因為會把待刪的建加回
//case 16/19 有大量的 insert 插入0...4999 再大量 remove
//Index 3358 out of bound for length 2141 表示在 remove 時，有殘留的index出錯
//if(list.size()>1 && i!=list.size()-1){...) 這樣就避開問題了！
/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 380/insert-delete-getrandom-o1.py
# insert() 重覆就 false
# remove() 不存在就 false
# getRandom() 就平均挑1個
# 速度都要O(1)，最多 2*10^5 次操作
# 本來我想用 set() 但不知怎麼亂數取出某個值
# 查到 SortedSet() 可用，但 remove() 及 [i] 要 O(log(n)) 不合規定。
# https://grantjenks.com/docs/sortedcontainers/sortedset.html
from sortedcontainers import SortedSet
class RandomizedSet:

    def __init__(self):
        self.pool = SortedSet()
        self.size = 0
    def insert(self, val: int) -> bool:
        if val in self.pool: return False
        self.pool.add(val)
        self.size += 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.pool: return False
        self.pool.remove(val) # 文件說這個耗時O(log(n))
        self.size -= 1 # 所以其實不合規定
        return True

    def getRandom(self) -> int:
        i = randrange(self.size)
        return self.pool[i] # 文件說這個耗時O(log(n))
        # 所以其實不合規定

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
